spring-cloud-gateway-oidc-tokenrelay/mock-hotels-service/src/main/java/com/jdriven/hotels/SecurityConfig.java

```diff
@@ -0,0 +1,80 @@
+package com.jdriven.hotels;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.springframework.boot.autoconfigure.security.oauth2.resource.OAuth2ResourceServerProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.oauth2.jwt.*;
+import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationConverter;
+
+@Configuration
+@EnableWebSecurity
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+	@Override
+	protected void configure(HttpSecurity http) throws Exception {
+		// Validate tokens through configured OpenID Provider
+		http.oauth2ResourceServer().jwt().jwtAuthenticationConverter(jwtAuthenticationConverter());
+		// Require authentication for all requests
+		http.authorizeRequests().anyRequest().authenticated();
+		// Allow showing pages within a frame
+		http.headers().frameOptions().sameOrigin();
+	}
+
+	private JwtAuthenticationConverter jwtAuthenticationConverter() {
+		JwtAuthenticationConverter jwtAuthenticationConverter = new JwtAuthenticationConverter();
+		// Convert realm_access.roles claims to granted authorities, for use in access decisions
+		jwtAuthenticationConverter.setJwtGrantedAuthoritiesConverter(new KeycloakRealmRoleConverter());
+		return jwtAuthenticationConverter;
+	}
+
+	@Bean
+	public JwtDecoder jwtDecoderByIssuerUri(OAuth2ResourceServerProperties properties) {
+		String issuerUri = properties.getJwt().getIssuerUri();
+		NimbusJwtDecoder jwtDecoder = (NimbusJwtDecoder) JwtDecoders.fromIssuerLocation(issuerUri);
+		// Use preferred_username from claims as authentication name, instead of UUID subject
+		jwtDecoder.setClaimSetConverter(new UsernameSubClaimAdapter());
+		return jwtDecoder;
+	}
+
+}
+
+// As per: https://docs.spring.io/spring-security/site/docs/5.2.x/reference/html5/#oauth2resourceserver-jwt-claimsetmapping-rename
+class UsernameSubClaimAdapter implements Converter<Map<String, Object>, Map<String, Object>> {
+
+	private final MappedJwtClaimSetConverter delegate = MappedJwtClaimSetConverter.withDefaults(Collections.emptyMap());
+
+	@Override
+	public Map<String, Object> convert(Map<String, Object> claims) {
```

Comment:
Goeie toevoeging dit om het username te parsen uit de JWT. Vind het wel raar/jammer dat die niet via een config property kan voor resourceserver zoals dat voor oauth clients wel kan om de user-info endpoint response te parsen.

Reply:
Heb even zitten kijken; beide config classes zitten in de Spring Boot nu; wel jammer dat het triviaal is bij de een, en best wat uitzoekwerk bij de ander inderdaad. Ik gok nog work in progress? Kan zo snel geen gerelateerde issues vinden..
Username extractie is in de resource server dus weer lastiger; andersom had ik het lastig van de week om de claims in de client (gateway) goed geextract te krijgen. Lijkt toch nog allemaal niet zo gebaande paden. Maargoed, de basis staat.